qa/tools/apex_algorithm_qa_tools/pytest_upload_assets.py

```diff
@@ -0,0 +1,172 @@
+
+
+def pytest_unconfigure(config):
+    if config.pluginmanager.hasplugin(_UPLOAD_ASSETS_PLUGIN_NAME):
+        config.pluginmanager.unregister(name=_UPLOAD_ASSETS_PLUGIN_NAME)
+
+
+class S3UploadPlugin:
+    def __init__(self, *, run_id: str | None = None, s3_client, bucket: str) -> None:
+        self.run_id = run_id or uuid.uuid4().hex
+        self.collected_assets: Dict[str, Path] | None = None
+        self.s3_client = s3_client
+        self.bucket = bucket
+        self.upload_stats = {"uploaded": 0}
+
+    def collect(self, path: Path, name: str):
+        """Collect assets to upload"""
+        assert self.collected_assets is not None, "No active collection of assets"
+        self.collected_assets[name] = path
+
+    def pytest_runtest_logstart(self, nodeid):
+        # Start new collection of assets for current test node
+        self.collected_assets = {}
+
+    def pytest_runtest_logreport(self, report: pytest.TestReport):
+        # TODO: option to upload on other outcome as well?
+        if report.when == "call" and report.outcome == "failed":
+            # TODO: what to do when upload fails?
+            uploaded = self._upload(nodeid=report.nodeid)
+            # TODO: report the uploaded assets somewhere (e.g. in user_properties or JSON report?)
+
+    def pytest_runtest_logfinish(self, nodeid):
+        # Reset collection of assets
+        self.collected_assets = None
+
+    def _upload(self, nodeid: str) -> Dict[str, str]:
+        assets = {}
+        for name, path in self.collected_assets.items():
+            safe_nodeid = re.sub(r"[^a-zA-Z0-9_.-]", "_", nodeid)
+            key = f"{self.run_id}!{safe_nodeid}!{name}"
+            # TODO: is this manual URL building correct and isn't there a boto utility for that?
+            url = f"{self.s3_client.meta.endpoint_url.rstrip('/')}/{self.bucket}/{key}"
+            _log.info(f"Uploading {path} to {url}")
+            self.s3_client.upload_file(
+                Filename=str(path),
+                Bucket=self.bucket,
+                Key=key,
+                # TODO: option to override ACL, or ExtraArgs in general?
+                ExtraArgs={"ACL": "public-read"},
+            )
+            assets[name] = url
+            self.upload_stats["uploaded"] += 1
+
+    def pytest_terminal_summary(self, terminalreporter):
+        terminalreporter.write_sep("-", f"`upload_assets` stats: {self.upload_stats}")
+
+
+@pytest.fixture
+def upload_assets(pytestconfig: pytest.Config, tmp_path) -> Callable:
+    """
+    Fixture to register a file (under `tmp_path`) for S3 upload
+    after the test failed. The fixture is a function that
+    can be called with one or more `Path` objects to upload.
+    """
+    uploader: S3UploadPlugin | None = pytestconfig.pluginmanager.get_plugin(
+        _UPLOAD_ASSETS_PLUGIN_NAME
+    )
+
+    if uploader:
+
+        def collect(*paths: Path):
+            for path in paths:
+                # TODO: option to make relative from other root
```

Comment:
TODO to avoid unnecessary "actual" component in key

Reply:
decided to leave this as-is. Better keep consistency and predictability than worrying about nice urls for now